### PR TITLE
kv (ticdc): fix kvClient reconnection downhill loop

### LINKED ISSUE #10239: TiCDC may incorrectly cancel gRPC streams to TiKV, causing the resolvedTs to become stuck
### What did you do?

We have found there could be many "stream to store closed" errors.
We think it's because of this logic in `cdc/kv/region_worker.go`:
```go
// checkShouldExit checks whether the region worker should exit, if exit return an error
func (w *regionWorker) checkShouldExit() error {
    empty := w.checkRegionStateEmpty()
    // If there is no region maintained by this region worker, exit it and
    // cancel the gRPC stream.
    if empty { /* It should be wrong because pending regions are not considered correctly. */
        w.cancelStream(time.Duration(0))
        return cerror.ErrRegionWorkerExit.GenWithStackByArgs()
    }
    return nil
}
```

### What did you expect to see?

_No response_

### What did you see instead?

Pending regions should be considered correctly in `checkShouldExit`.

### Versions of the cluster

6.5.x, 7.1.x.

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #10239 

### What is changed and how it works?

1. Add an ID to the `eventFeedStream` struct to identify a stream and prevent it from being deleted unexpectedly.
2. Bind the cancel function of a gRPC stream to its `eventFeedStream` to prevent the stream from being canceled unexpectedly.
3. Reduce the number of calls to `s.deleteStream` to only once to prevent the stream from being deleted unexpectedly.

#### Tests 

 - Manual test (add detailed scripts or steps below)

1. Add hard code `time.sleep(5 * time.second)` before calling `s.deleteStream` in both unfixed and fixed version of code. 
2. Deploy a TiDB cluster with 3 TiKV nodes and 1 TiCDC server, create a changefeed. 
3. Patch the changed cdc binary, and restart a TiKV node randomly and observe the changefeed's lag. 

unfixed cdc:
img_v3_027f_9b5e20b7-2aab-44b7-8884-fc29d2797beg

fixed cdc:
image
 
From the above graphs, it is evident that in the unfixed CDC, the lag of resolvedTs can exceed 12 minutes when a TiKV node is restarted. However, in the fixed CDC, the increase in resolvedTs is limited to a maximum of 35 seconds. This demonstrates the effectiveness of the fix.

Moreover, when the hard-coded `time.sleep(5 * time.second)` is removed and the fixed version of CDC is tested again, the lag becomes even smaller:
image

### Release note 

```release-note
Fix a bug in kv client that could cause an increase in changefeed lag when TiKV is restarted.
```

### CODE REVIEW

File: cdc/kv/client.go [hunk 1]
```diff
@@ -752,14 +769,16 @@ func (s *eventFeedSession) requestRegionToStore(
 					zap.String("store", storeAddr),
 					zap.Error(err))
 			}
-			// Delete the stream from the map so that the next time the store is accessed, the stream will be
-			// re-established.
-			s.deleteStream(storeAddr)
+			// TODO(dongmen): remove this line after testing.
+			time.Sleep(time.Second * 5)
```
Comment (hicqu) [inline cdc/kv/client.go:773]: Do tests still need this? We can use a failpoint.

File: cdc/kv/region_worker.go [hunk 2]
```diff
@@ -595,21 +600,18 @@ func (w *regionWorker) checkErrorReconnect(err error) error {
...
-			zap.String("namespace", w.session.client.changefeed.Namespace),
-			zap.String("changefeed", w.session.client.changefeed.ID))
-	}
+	// cancel the stream to make strem.Recv returns a context cancel error
+	// This will make the receiveFromStream goroutine exit and the stream can
+	// be re-established by the caller.
+	// Note: use context cancel is the only way to terminate a gRPC stream.
+
+	// TODO(dongmen): remove this line after testing.
+	time.Sleep(time.Second * 5)
```
Comment (hicqu) [inline cdc/kv/region_worker.go:609]: Needs updates.

**Comment by asddongmen** [comment]
I found some others problem in kvClient when I testing this PR. Further investigation is needed.

File: cdc/kv/client.go [hunk 3]
```diff
@@ -232,7 +242,12 @@ func NewCDCClient(
 	return c
 }
 
-func (c *CDCClient) newStream(ctx context.Context, addr string, storeID uint64) (stream *eventFeedStream, newStreamErr error) {
+func (c *CDCClient) newStream(
+	ctx context.Context,
+	cancel context.CancelFunc,
```
Comment (3AceShowHand) [inline cdc/kv/client.go:247]: pass the `cancel` as a function argument looks so weired.

Comment (asddongmen) [reply cdc/kv/client.go:247]: I have removed it.

**Comment by asddongmen** [comment]
> I found some other problems in kvClient when I testing this PR. Further investigation is needed.

I add a `streamAlterInterval` to prevent a stream of a same store being created/deleted frequently.

**Review by hicqu — APPROVED** [review]

**Review by 3AceShowHand — APPROVED** [review]

**Pull request merged**